### DISCUSSION TITLE
Code Output Comparison and Correction Request

### DISCUSSION THREAD

**Discussion by yusuforzibekov** [community/community#75952] [Programming Help]
### Body

```JAVA
class Solution {

    /**
     * @param n the given positive number
     * @return the square matrix of order n according to the given pattern
     */
    public static int[][] create(int n) {
        int[][] matrix = new int[n][n];

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (i == 0 || i == n - 1 || j == 0 || j == n - 1) {
                    matrix[i][j] = 1;
                }
            }
        }

        return matrix;
    }

    // Helper method to print the matrix
    public static void printMatrix(int[][] matrix) {
        for (int i = 0; i < matrix.length; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                System.out.print(matrix[i][j] + " ");
            }
            System.out.println();
        }
        System.out.println("\n");
    }

    public static void main(String[] args) {
        // Test cases
        int[][] result1 = create(5);
        printMatrix(result1);

        int[][] result2 = create(6);
        printMatrix(result2);

        int[][] result3 = create(1);
        printMatrix(result3);

        int[][] result4 = create(2);
        printMatrix(result4);
    }
}
```

Hi everyone,

When I test the code above, it outputs like this:

![image](https://github.com/community/community/assets/75078482/7c7109f1-2dd8-47a6-8b2c-11147cb6f122)

However, there are some issues in my code for outputting the correct version of the visibility. It should output like this:

![image](https://github.com/community/community/assets/75078482/e7d779b9-e70c-49dc-9225-5528b1a4cbee)

And to others, it is correctly outputted.

So how can i output it correctly. Any ideas?

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by ghost** [comment]
Hello @yusuforzibekov 

If I understand your task correctlly this is how i fixed it:
in your **create** function where you create matrix and set 1 in the if statement you need to change condition of the if block. Curentlly you use 
`i == 0 || i == n - 1 || j == 0 || j == n - 1` condition to set 1 in the matrix which sets 1 on the boaders of the matrix. however this is not what you need. i changed your if block condition from that to this: 
```js
if(
    i <= n/2 && ( i >= j || n - i - 1 <= j ) || 
    i >= n/2 && ( n - i > j || n - j <= n - i )
){
    matrix[i][j] = 1;
}
```
changing this part will fix the problem.

**However** i think there can be better approach to solve this and simmilar problems. Its called **Recursion**. you can define base case for your problem and then call your function for smaller problem. I will not talk any more about recursion but if you have never heard it before check it [there](https://blog.udemy.com/java-recursion/?utm_source=adwords&utm_medium=udemyads&utm_campaign=DSA_Catchall_la.EN_cc.ROW&utm_content=deal4584&utm_term=_._ag_88010211481_._ad_535397282061_._kw__._de_c_._dm__._pl__._ti_dsa-449490803887_._li_1007469_._pd__._&matchtype=&gad_source=1&gclid=Cj0KCQiA3uGqBhDdARIsAFeJ5r1tYLNPBXltqC-tdG8Z5si162akmo9Tv3Oupij7ZkOPGIDV6N9VSYEaAuAsEALw_wcB). worth to learn principles of it.

I have not implemented recurion solution for this problem but there are basic steps how you can do it by yourself:

- define base case. there base case solution is matrix of 1X1 or 2X2 (depending on how you implement next steps you will have to define either 1X1 or both of them) and for both cases you should return (not return but to fill whole 1X1 or 2X2 matrix with number **1**)

- come up with step to make your problem smaller after each step that will lead to base case that you have defined. for your case such a step is to fill left and right boarder with number 1 and then call your function (function which does recursion. where you defined base case) with smaller matrix. in your case smaller matrix is sub-matrix of original one (when you fill boarders with 1 then you should take inside matrix and recall your function on that). for example if your matrix is 5X5 after firs time you fill boarder your smaller sub-matrix will be 3X3 (after "cutting" boarders from all four side on 5X5)

- recursion solution repeats itself and finally it will end up with base case and recursion will be stopped. **wholaa, problem is solved there**. 

**NOTE:** 
- Recursion is not a toy to play with, if you do define base case or you define it but it does not covers all the possible variants you will end up with infinit loop and you will get **stack overflow**  error (in best case). in worst case your machine will freeze. for example if you include only 1X1 matrix in your base case and will not keep this in mind during next steps after 6X6 -> 4X4 -> 2X2 -> 0X0 -> ... (of corse negative length matrix does not exists but your recursion will not end up with base case and you will face mantioned problems)

**Comment by queenofcorgis** [comment]
Thanks for posting in the GitHub Community, @yusuforzibekov ! 

We're happy you're here. You are more likely to get a useful response if you are posting your question in the applicable category, the Projects and Issues category is solely related to conversations around the GitHub products Issues and Projects.

We’ve moved your post to our Programming Help 🧑‍💻 category, which is more appropriate for this type of discussion.

Please review our [guidelines](https://github.com/orgs/community/discussions/51384) about the Programming Help category for more information.
